openmicroscopy/openmicroscopy
# Api pagination

# What this PR does

Updates pagination to use ```?offset=200``` to show page 2 of results instead of ```?page=2```.
We also count the total number of objects that would be returned by all "list objects" queries and return this in a new ```"meta"``` json object, along with other pagination info (current limit, offset & max limit):
E.g. ```/images/```
```
"meta": {
    "totalCount": 16739,
    "limit": 200,
    "maxLimit": 500,
    "offset": 0
},
```
This PR also adds new config settings for the default api "limit" and the "max limit" in a new ```api_settings.py```. Other api settings have also been moved to this file.

# Testing this PR

1. Go to any top-level object listing, e.g. /projects/ or /images/ and check for ```"meta"``` in json.
    - Check that 'totalCount' is correct (same as in clients)
2. Filter E.g. filter images by Dataset / orphaned and check that totalCount is correct.
3. Try using ```?offset=100``` and / or ```?limit=50```. Check that these values are shown in "meta" (and that they do pagination as expected)
4. Test the max limit by trying a limit above 500. E.g. use ```?limit=1000```. Should see that the actual limit used in the query is restricted to the max_limit (500).
5. Configure the default limit ```bin/omero config set omero.web.api.limit 300```
6. Configure the max limit ```bin/omero config set omero.web.api.max_limit 1000```
7. Check that these are used correctly in the scenarios above.

# Related reading

Discussed at web meeting [2017-01-26](https://docs.google.com/document/d/18ep65grhBURv6CcZLGVmvKQvP3RjfuyIhE288AAuFTM/edit#heading=h.9u06hgeemoyv) and [2017-02-08](https://docs.google.com/document/d/18ep65grhBURv6CcZLGVmvKQvP3RjfuyIhE288AAuFTM/edit#heading=h.3ekyh84japrw)
